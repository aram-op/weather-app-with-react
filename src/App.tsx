import React, {useState} from 'react';
import './App.css';
import Forecast from './components/Forecast';
import ChangeUnitButton from './components/ChangeUnitButton';
import CurrentWeather from './components/CurrentWeather';
import {Unit} from './models/unit.model';
import ErrorMessage from './components/ErrorMessage';

function App() {
    const [unit, setUnit] = useState<Unit>('metric');
    const [error, setError] = useState<any>(null);

    function handleUnitChange() {
        if (unit === 'metric') {
            setUnit('imperial');
        } else {
            setUnit('metric');
        }
    }

    function handleError(error: Error) {
        console.error(error);
        setError(error);
    }

    if (error) return (
        <ErrorMessage message="Error fetching weather data :("/>
    );

    return (
        <div className="container">
            <ChangeUnitButton onButtonClicked={handleUnitChange}/>
            <CurrentWeather unit={unit} key={`current weather ${unit}`} onError={handleError}/>
            <Forecast unit={unit} key={`forecast ${unit}`} onError={handleError}/>
        </div>
    );
}

export default App;
