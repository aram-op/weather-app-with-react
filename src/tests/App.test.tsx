import React from 'react';
import {Unit} from '../models/unit.model';
import {fireEvent, render, screen} from '@testing-library/react';
import App from '../App';

jest.mock('../components/Forecast', () => () => <div>Forecast</div>);
jest.mock('../components/ChangeUnitButton', () => ({onUnitChange}: { onUnitChange: () => void }) => (
    <button onClick={onUnitChange}>Change Unit</button>
));
jest.mock('../components/CurrentWeather', () => ({unit, onError}: { unit: Unit, onError: () => void }) => (
    <>
        <div>Current Weather, unit = {unit}</div>
        <button onClick={onError}>Trigger error</button>
    </>
));
jest.mock('../components/ErrorMessage', () => ({message}: { message: string }) => <h2>{message}</h2>);


describe('App component', () => {
    it('should render and contain the UnitChangeButton, CurrentWeather and Forecast components', () => {
        render(<App/>);

        expect(screen.getByText('Change Unit')).toBeInTheDocument();
        expect(screen.getByText('Forecast')).toBeInTheDocument();
        expect(screen.getByText('Current Weather, unit = metric')).toBeInTheDocument();
    });

    it('should change the unit when the button is clicked', () => {
        render(<App/>);

        expect(screen.getByText('Current Weather, unit = metric')).toBeInTheDocument();

        fireEvent.click(screen.getByText('Change Unit'));

        expect(screen.getByText('Current Weather, unit = imperial')).toBeInTheDocument();
    });

    it('should render the error message component when there is an error', () => {
        render(<App/>);

        fireEvent.click(screen.getByText('Trigger error'));

        expect(screen.getByText('Error fetching weather data :(')).toBeInTheDocument();
    });

    it('should match snapshot', () => {
        const {asFragment} = render(<App/>);

        expect(asFragment()).toMatchSnapshot();
    });
});
