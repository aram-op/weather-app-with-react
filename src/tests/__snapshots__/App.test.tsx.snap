// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="container"
  >
    <button>
      Change Unit
    </button>
    <div>
      Current Weather, unit = metric
    </div>
    <button>
      Trigger error
    </button>
    <div>
      Forecast
    </div>
  </div>
</DocumentFragment>
`;

exports[`renders ChangeUnitButton, CurrentWeather and Forecast components 1`] = `
<div>
  <div
    class="container"
  >
    <button
      class="unit-change-btn"
    >
      Change Unit
    </button>
    <div
      class="forecast-container"
    />
  </div>
</div>
`;
